orthanc/orthanc-anon/plugin/pixl.py

```
@@ -219,42 +216,151 @@ def OnHeartBeat(output, uri, **request) -> Any:  # noqa: ARG001
+    Send(study_id=anonymised_study_resource_id)
+
+
+def _get_existing_study(study_uid: str) -> Optional[str]:
+    """
+    Get the resource ID for an existing study based on its StudyInstanceUID.
+
+    Returns None if there are no resources with the given StudyInstanceUID.
+    Returns the resource ID if there is a single resource with the given StudyInstanceUID.
+    Returns None if there are multiple resources with the given StudyInstanceUID and deletes
+    the studies.
+    """
+    data = json.dumps(
+        {
+            "Level": "Study",
+            "Query": {
+                "StudyInstanceUID": study_uid,
+            },
+        }
+    )
+    study_resource_ids = json.loads(orthanc.RestApiPost("/tools/find", data))
+    if not study_resource_ids:
+        logger.info(f"No study found with StudyInstanceUID {study_uid}")
+        return None
+
+    if len(study_resource_ids) == 1:
+        return study_resource_ids[0]
+
+    logger.info(
+        f"{len(study_resource_ids)} found with StudyInstanceUID {study_uid}. Deleting studies."
+    )
+
+    return None
+
+
+def _anonymise_study_instances(zipped_study: ZipFile, study_uid: str) -> tuple[list[bytes], str]:
+    """
+    Iterate over all instances and anonymise them.
+
+    Return a list of the bytes of anonymised instances, and the anonymised StudyInstanceUID.
+    """
+    anonymised_instances_bytes = []
+    for file_info in zipped_study.infolist():
+        with zipped_study.open(file_info) as file:
+            logger.debug("Reading file {}", file)
+            try:
+                dataset = dcmread(file)
+            except pydicom.errors.InvalidDicomError:
+                logger.warning("Failed to read file {}.", file)
+                raise
+
+            try:
+                logger.debug("Anonymising file: {}", file)
+                anonymised_instances_bytes.append(_anonymise_dicom_instance(dataset))
+            except PixlDiscardError:
+                pass
```

Am I right in thinking that if a `PIXLDiscardError` is raised, we should just skip this instance? If any other is raised, we should raise an error and not uploaded the anonymised instanes?

---

I think that's right, but I think `PIXLDiscardError` is raised when there is an error on querying the database, so we should change that to another exception type

---

ah okay, good idea, I'll check where else `PIXLDiscardError` is used too

---

I've added a `PixlSkipInstanceError` that's raised if we should skip an instance based on the project config